feat(ci): Add CI pipeline and pre-commit hooks

thedgarg31 opened the pull request:
✅ Clean CI Pipeline & Pre-commit Hook Integration
This PR introduces a clean, focused implementation of the requested CI and pre-commit infrastructure, as per Anirudh’s instructions.

🔧 What's Included:
Pre-commit Hooks (Linux/macOS & Windows)

scripts/install-hooks.sh and install-hooks.bat

Runs clang-format & clang-tidy on staged C++ files

CI Build Workflows

.github/workflows/build.yml

Separate jobs for Linux, macOS, and Windows (uses Ninja)

Config: BUILD_TESTS=OFF, BUILD_EXAMPLES=OFF

Docs

DEVELOPMENT.md with setup and usage instructions

🚫 What's Not Touched:
No changes to source, test, deprecated files, or CMakeLists.txt

✅ The branch is ready for testing: feat/ci-pipeline-clean
Let's verify the Windows build and proceed to finalize!

thedgarg31 committed:
feat: Add CI pipeline and pre-commit hooks

- Add build workflow for Linux, macOS, and Windows
- Add pre-commit hook installation scripts for clang-format and clang-tidy
- Add development setup documentation
- Focus only on build infrastructure and linting setup

Scope:
✅ Pre-commit hooks for linting and tidying
✅ Build workflows only
❌ No source code modifications
❌ No test modifications
❌ No deprecated file changes

SharonIV0x86 commented:
@thedgarg31 Alright, the workflows seem to work and the formatting also works on linux with commits, but i am not sure about the windows.
You need to rename the ``DEVELOPMENT.md`` to ``development.md`` and move this file to the ``/docs`` folder.

SharonIV0x86 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,71 @@
+        cmake-version: '3.20'
+
+    - name: Configure CMake
+      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=OFF
```
```suggestion
      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

SharonIV0x86 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,71 @@
+        cmake-version: '3.20'
+
+    - name: Configure CMake
+      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=OFF
```
```suggestion
      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

SharonIV0x86 commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,71 @@
+      uses: ilammy/msvc-dev-cmd@v1
+
+    - name: Configure CMake
+      run: cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=OFF
```
```suggestion
      run: cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

SharonIV0x86 reviewed (changes requested):
Need changes.

SharonIV0x86 committed:
Deprecated vis engine.

thedgarg31 committed:
Update .github/workflows/build.yml

Co-authored-by: Anirudh Lakhanpal <91114837+SharonIV0x86@users.noreply.github.com>

thedgarg31 committed:
Update .github/workflows/build.yml

Co-authored-by: Anirudh Lakhanpal <91114837+SharonIV0x86@users.noreply.github.com>

thedgarg31 committed:
Update .github/workflows/build.yml

Co-authored-by: Anirudh Lakhanpal <91114837+SharonIV0x86@users.noreply.github.com>

thedgarg31 committed:
docs: Move DEVELOPMENT.md to docs/development.md

- Rename DEVELOPMENT.md to development.md (lowercase)
- Move file from root to docs/ folder as requested in PR review

thedgarg31 commented:
I’ve implemented all the requested changes, including renaming and moving DEVELOPMENT.md to docs/development.md. However, after the updates, the CI workflows for macOS, Linux, and Windows are now failing. Could you please help review the build logs so we can identify and fix the root cause?

SharonIV0x86 commented:
@thedgarg31 All the workflows seem to work as expected. Check twice.

thedgarg31 commented:
"Thanks for confirming the workflows are working fine. I’ve double-checked and all checks are passing now. Ready for merge from my side ✅"

SharonIV0x86 reviewed (approved):
Approved.
Will be merged in sometime. Till then if you want you can work on some new issues.

thedgarg31 commented:
Sure.Thanku

SharonIV0x86 commented:
@thedgarg31 
To assign your points for GSSOC, I’ll need your full name (or the exact name you registered with for GSSOC). Please ensure the information is accurate, as incorrect details may result in your points not being credited.

thedgarg31 commented:
Ya Sure .
I registered with my full name - DAKSH GARG 
Thanks

SharonIV0x86 commented:
@thedgarg31 Thank you for contributing, your PR has been merged.
Kindly ⭐ the repository as it shows appreciation to the maintainers.

thedgarg31 commented:
Done

thedgarg31 commented:
Thanks for merging my PR and assigning the GSSoC points. Could you please let me know when they will be reflected on the leaderboard and where I can check them?

SharonIV0x86 commented:
> Thanks for merging my PR and assigning the GSSoC points. Could you please let me know when they will be reflected on the leaderboard and where I can check them?

There is a lot of confusion with the official leaderboard and the whole points assigning process is manual, like PAs have to manually keep track of the points in a google sheet and submit that sheet to the managers whenever requested, so its nothing in our hands it can take a week or so to get your points reflected on the leaderboard once it is live but i am not sure.

SharonIV0x86 commented:
@thedgarg31 You can check your points here
https://gssoc25-leaderboard.vercel.app/download
search by the project.

thedgarg31 commented:
Thanku so much
